data-analytics/dataflow-elasticsearch-indexer/ElasticIndexer/src/main/java/com/google/cloud/pso/common/FailSafeValidate.java

```diff
@@ -0,0 +1,147 @@
+ * 2. The well formed json document contains key element specified as a json pointer
+ * The PTransform returns a {@link PCollectionTuple} tagged with the {@link TupleTag}
+ * provided.
+ */
+@AutoValue
+public abstract class FailSafeValidate
+        extends PTransform<PCollection<String>, PCollectionTuple> {
+
+    abstract TupleTag<KV<String, String>> successTag();
+    abstract TupleTag<ErrorMessage> failureTag();
+    abstract ObjectReader objectReader();
+    abstract ExtractKeyFn extractKeyFn();
+
+    public static Builder newBuilder() {
+        return new AutoValue_FailSafeValidate.Builder()
+                .setObjectReader((new ObjectMapper()).reader());
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+        abstract Builder setSuccessTag(TupleTag<KV<String, String>> successTag);
+
+        abstract Builder setFailureTag(TupleTag<ErrorMessage> failureTag);
+
+        abstract Builder setObjectReader(ObjectReader objectReader);
+
+        abstract Builder setExtractKeyFn(ExtractKeyFn extractKeyFn);
+
+        public abstract FailSafeValidate build();
+
+
+        public Builder withKeyPath(String keyPath) {
+            checkArgument(keyPath != null,
+                    "withKeyPath(keyPath) called with null value.");
+            try {
+                JsonPointer.compile(keyPath);
+            } catch (IllegalArgumentException e) {
+                throw new RuntimeException(e);
+            }
+            return setExtractKeyFn(new ExtractKeyFn(keyPath));
+        }
+
+        public Builder withSuccessTag(TupleTag<KV<String, String>> successTag) {
+            checkArgument(successTag != null,
+                    "withSuccessTag(successTag) called with null value.");
+            return setSuccessTag(successTag);
+        }
+
+        public Builder withFailureTag(TupleTag<ErrorMessage> failureTag) {
+            checkArgument(failureTag != null,
+                    "withFailureTag(failureTag) called with null value.");
+            return setFailureTag(failureTag);
+        }
+    }
+
+
+    @Override
+    public PCollectionTuple expand(PCollection<String> input) {
+        return input.apply("ValidateAndTagMessages",
+                ParDo.of(
+                        new DoFn<String, KV<String, String>>() {
+                            @ProcessElement
+                            public void processElement(ProcessContext context) {
+                                String input = context.element();
+
+                                try {
+                                    JsonParser parser = objectReader().getFactory().createParser(input);
+                                    // Parse the json to validate it. An IOException is thrown if
+                                    // it encounters a parsing error.
+                                    while(parser.nextToken() != null) {}
+                                    JsonNode jsonNode = objectReader().readTree(input);
```

Comment:
Seems like this would convert the JSON twice. Do we need the above while loop if we are reading the tree subsequently?